derfinderExample
================

# Introduction

Example on how to use the __[derfinder2](https://github.com/lcolladotor/derfinder2)__ package using a subset of the data data from the 2010 papers by [Montgomery et al](http://www.ncbi.nlm.nih.gov/pubmed?term=20220756) and [Pickerell et al](http://www.ncbi.nlm.nih.gov/pubmed?term=20220758).

The goal of the repository is to show how to use __derfinder2__ with your own data and to help you get started with actually running an full scale analysis with __derfinder2__. 

For documentation related to __derfinder2__ itself check [it's GitHub repository](https://github.com/lcolladotor/derfinder2) and help files from R `help(package="derfinder2")`.


# Overview

## Setup

A set of HapMap samples was downloaded to [reads](https://github.com/lcolladotor/derfinderExample/tree/master/reads) using information from the HapMap pedigrees in [info](https://github.com/lcolladotor/derfinderExample/tree/master/info). The iGenome human reference needed to run __TopHat__ was downloaded in [ref](https://github.com/lcolladotor/derfinderExample/tree/master/ref). Then the reads were aligned using __TopHat__ in [thout](https://github.com/lcolladotor/derfinderExample/tree/master/thout) and this is the more computer intensive step in the whole analysis.

## derfinder2 setup

The coverage information was processed using __derfinder2__ and prepared for downstream analyses. 

First, the un-filtered coverage was saved in [derCoverageInfo](https://github.com/lcolladotor/derfinderExample/tree/master/derCoverageInfo) as `fullCov.Rdata`. This can be useful later on for making coverage plots to evaluate whether visually whether the method is picking up sensible regions. 

Second, the filtered coverage was saved in [derCoverageInfo](https://github.com/lcolladotor/derfinderExample/tree/master/derCoverageInfo) separately for each chromosome in `chr*CovInfo.Rdata`.

## derfinder2 analysis

The main steps of the __derfinder2__ analysis pipeline are carried out in [derAnalysis/derfinder2-analysis.R](https://github.com/lcolladotor/derfinderExample/blob/master/derAnalysis/derfinder2-analysis.R) located in the [derAnalysis](https://github.com/lcolladotor/derfinderExample/tree/master/derAnalysis) folder. The richness of your data set (coverage depth and number of samples) and the number of cores needed for wallclock speed purposes will greatly determine how much RAM __derfinder2__ will use. On this example data set, chromosome 1 took 20 mins 40 secs and 12.65 GB of RAM with 2 cores. With another richer data set, the same chromosome took 1 hr 14 mins and 128 GB of RAM using 8 cores. Note that we could have exchanged wallclock time for RAM if we needed to.

Once `derAnalysis/derfinder2-analysis.R` has completed its work, [derAnalysis/derfinder2-merge.R](https://github.com/lcolladotor/derfinderExample/blob/master/derAnalysis/derfinder2-merge.R) is used to merge the results from running __derfinder2__ on each chromosome.

## HTML report

An HTML report to carry out an exploration of the results can be generated by using [derAnalysis/basicExploration.Rmd](https://github.com/lcolladotor/derfinderExample/blob/master/derAnalysis/basicExploration.Rmd) via [derAnalysis/basicExplore-run1.sh](https://github.com/lcolladotor/derfinderExample/blob/master/derAnalysis/basicExplore-run1.sh) or similar shell scripts. Reports range in the 30 to 40 MB size and take 1-2 hrs to generate (making the coverage plots takes time using __ggbio__). More cores (and RAM) are needed only if __derfinder2__ generated millions of null statistics, in which case you probably need to use a more stringent cutoff when running __calculatePvalues()__.

Please download this repository and open [derAnalysis/derEx-basicExplore-run1/basicExplore.html](https://github.com/lcolladotor/derfinderExample/blob/master/derAnalysis/derEx-basicExplore-run1/basicExplore.html) (it needs [derAnalysis/derEx-basicExplore-run1/figure](https://github.com/lcolladotor/derfinderExample/tree/master/derAnalysis/derEx-basicExplore-run1/figure) to open properly) to view the report generated for this example. If you do not want to download the full repository (30-40 MBs shouldn't bee too much!) you can view the plots from the report in [derAnalysis/derEx-basicExplore-run1/figure](https://github.com/lcolladotor/derfinderExample/tree/master/derAnalysis/derEx-basicExplore-run1/figure).


# Installation instructions

Get R 3.0.1 or newer from [CRAN](http://cran.r-project.org/).

```S
## If needed
install.packages("devtools")

## Pre-requisites from CRAN
install.packages(c("knitr", "Rcpp", "RcppArmadillo", "ggplot2", "reshape2", "plyr", 
	"microbenchmark", "gridExtra", "data.table", "knitr", "knitcitations",
	"xtable", "RColorBrewer", "scales"))

## Pre-requisites from Bioconductor
source("http://bioconductor.org/biocLite.R")
biocLite(c("IRanges", "GenomicRanges", "Rsamtools", "bumphunter", "biovizBase", "ggbio", "qvalue",
	 "TxDb.Hsapiens.UCSC.hg19.knownGene", "AnnotationDbi", "GenomicFeatures"))

## GitHub dependencies
library(devtools)
install_github("rCharts", "ramnathv", ref="dev")
install_github(username="jimhester", repo="knitrBootstrap")

## derfinder2 itself
library(devtools)
install_github("derfinder2", "lcolladotor")

## Other packages used in this example that are not derfinder2 pre-reqs
install.packages("getopt")
biocLite("SRAdb")
```

Note that the current Bioconductor release version of __bumphunter__ for R 3.0.1 is a few versions before the one required by __derfinder2__. The version needed can be installed manually from http://bioconductor.org/packages/2.13/bioc/html/bumphunter.html You can download the source or other binaries.

# Citation

Below is the citation output from using `citation("derfinder2")` in R. Please run this yourself to check for any updates on how to cite __derfinder2__.

---

To cite package __derfinder2__ in publications use:

Leonardo Collado-Torres, Alyssa Frazee, Andrew Jaffe and Jeffrey Leek (2013). derfinder2: Fast differential expression analysis of RNA-seq data at base-pair resolution. R package version 0.0.24. https://github.com/lcolladotor/derfinder2

A BibTeX entry for LaTeX users is

@Manual{, title = {derfinder2: Fast differential expression analysis of RNA-seq data at base-pair resolution}, author = {Leonardo Collado-Torres and Alyssa Frazee and Andrew Jaffe and Jeffrey Leek}, year = {2013}, note = {R package version 0.0.24}, url = {https://github.com/lcolladotor/derfinder2}, }
